File: mobile-app/lib/services/pdf_service.dart
import 'package:barcode/barcode.dart' as bc;
import 'package:flutter/services.dart' show rootBundle;
import 'package:pdf/pdf.dart';
import 'package:pdf/widgets.dart' as pw;
import 'package:intl/intl.dart';
import '../models/invoice.dart';
import '../models/order.dart';

class PdfService {
  static final PdfService _instance = PdfService._internal();
  factory PdfService() => _instance;
  PdfService._internal();

  static const _green = PdfColor.fromInt(0xFF1F6F4A); // Saudi flag green
  static const _maroon = PdfColor.fromInt(0xFF8D1B3D);
  static const _gold = PdfColor.fromInt(0xFFC9A84C);
  static const _gray = PdfColor.fromInt(0xFF6B7280);
  static const _lightGray = PdfColor.fromInt(0xFFF3F4F6);
  static const _whiteLight = PdfColor(1, 1, 1, 0.75);

  // Strip non-ASCII chars — PDF Type1 fonts are Latin-1 only; Arabic glyphs render blank
  String _safeCurrency(String? sym) {
    if (sym == null || sym.isEmpty) return 'SAR';
    final ascii = sym.replaceAll(RegExp(r'[^\x00-\x7F]'), '').trim();
    return ascii.isEmpty ? 'SAR' : ascii;
  }

  String _fmt(NumberFormat fmt, num? value) => fmt.format(value ?? 0);

  pw.Font? _arabicFont;
  pw.Font? _arabicBold;

  /// Arabic is mandatory on a Saudi tax invoice, and PDF core fonts cannot
  /// render it — Amiri is bundled with the app and loaded once.
  Future<void> _loadArabicFonts() async {
    if (_arabicFont != null) return;
    try {
      _arabicFont = pw.Font.ttf(await rootBundle.load('assets/fonts/Amiri-Regular.ttf'));
      _arabicBold = pw.Font.ttf(await rootBundle.load('assets/fonts/Amiri-Bold.ttf'));
    } catch (_) {
      // Font missing — the invoice still prints, in English only
    }
  }

  pw.Widget _ar(String text, {double size = 9, bool bold = false, PdfColor color = PdfColors.black}) {
    if (_arabicFont == null) return pw.SizedBox();
    return pw.Directionality(
      textDirection: pw.TextDirection.rtl,
      child: pw.Text(text, style: pw.TextStyle(
        font: bold ? _arabicBold : _arabicFont,
        fontSize: size, color: color,
        fontWeight: bold ? pw.FontWeight.bold : pw.FontWeight.normal,
      )),
    );
  }

  Future<List<int>> generateInvoicePdf({
    required Invoice invoice,
    required List<OrderItem> items,
    Map<String, String>? settings,
  }) async {
    await _loadArabicFonts();

    final doc = pw.Document();
    final currency = _safeCurrency(settings?['currency']);
    final companyName = settings?['company_name'] ?? 'Dubyx Trading Est.';
    final companyNameAr = settings?['company_name_ar'] ?? '';
    final companyAddress = settings?['company_address'] ?? 'Riyadh, Saudi Arabia';
    final companyPhone = settings?['company_phone'] ?? '';
    final companyEmail = settings?['company_email'] ?? '';
    final sellerVat = settings?['company_vat_number'] ?? '';
    final sellerCr = settings?['company_cr_number'] ?? '';
    final vatRate = settings?['default_tax_rate'] ?? '15';
    // Force 'en' locale — device Arabic locale would emit non-ASCII numerals that PDF Type1 fonts can't render
    final fmt = NumberFormat('#,##0.00', 'en');

    doc.addPage(pw.Page(
      pageFormat: PdfPageFormat.a4,
      margin: const pw.EdgeInsets.all(32),
      build: (ctx) => pw.Column(
        crossAxisAlignment: pw.CrossAxisAlignment.start,
        children: [
          // Header
          pw.Container(
            padding: const pw.EdgeInsets.all(20),
            decoration: const pw.BoxDecoration(color: _green, borderRadius: pw.BorderRadius.all(pw.Radius.circular(12))),
            child: pw.Row(
              mainAxisAlignment: pw.MainAxisAlignment.spaceBetween,
              children: [
                pw.Column(crossAxisAlignment: pw.CrossAxisAlignment.start, children: [
                  pw.Text(companyName, style: pw.TextStyle(color: PdfColors.white, fontSize: 22, fontWeight: pw.FontWeight.bold)),
                  pw.SizedBox(height: 4),
                  if (companyNameAr.isNotEmpty) _ar(companyNameAr, size: 12, bold: true, color: PdfColors.white),
                  pw.Text(companyAddress, style: const pw.TextStyle(color: _whiteLight, fontSize: 9)),
                  pw.Text('$companyPhone  |  $companyEmail', style: const pw.TextStyle(color: _whiteLight, fontSize: 9)),
                  if (sellerVat.isNotEmpty) pw.Text('VAT No: $sellerVat', style: const pw.TextStyle(color: _whiteLight, fontSize: 9)),
                  if (sellerCr.isNotEmpty) pw.Text('CR No: $sellerCr', style: const pw.TextStyle(color: _whiteLight, fontSize: 9)),
                ]),
                pw.Column(crossAxisAlignment: pw.CrossAxisAlignment.end, children: [
                  pw.Text(invoice.titleEn, style: pw.TextStyle(color: _gold, fontSize: 15, fontWeight: pw.FontWeight.bold)),
                  _ar(invoice.titleAr, size: 14, bold: true, color: PdfColors.white),
                  pw.SizedBox(height: 2),
                  pw.Text(invoice.invoiceNumber, style: const pw.TextStyle(color: PdfColors.white, fontSize: 11)),
                  if (invoice.uuid != null)
                    pw.Text('UUID: ${invoice.uuid}', style: const pw.TextStyle(color: _whiteLight, fontSize: 6)),
                ]),
              ],
            ),
          ),
          pw.SizedBox(height: 20),

          // Bill To / Invoice Info
          pw.Row(crossAxisAlignment: pw.CrossAxisAlignment.start, children: [
            pw.Expanded(child: pw.Column(crossAxisAlignment: pw.CrossAxisAlignment.start, children: [
              pw.Text('BILL TO', style: pw.TextStyle(color: _gray, fontSize: 8, fontWeight: pw.FontWeight.bold)),
              pw.SizedBox(height: 4),
              pw.Text(invoice.customerName, style: pw.TextStyle(fontSize: 13, fontWeight: pw.FontWeight.bold)),
              pw.Text(invoice.customerPhone, style: const pw.TextStyle(color: PdfColors.grey600, fontSize: 10)),
              pw.Text(invoice.customerAddress, style: const pw.TextStyle(color: PdfColors.grey600, fontSize: 10)),
              if (invoice.customerVatNumber != null)
                pw.Text('VAT No: ${invoice.customerVatNumber}', style: const pw.TextStyle(color: PdfColors.grey600, fontSize: 10)),
            ])),
            pw.SizedBox(width: 40),
            pw.Expanded(child: pw.Column(crossAxisAlignment: pw.CrossAxisAlignment.end, children: [
              _infoRow('Date:', DateFormat('dd MMM yyyy').format(invoice.createdAt)),
              _infoRow('Due Date:', invoice.dueDate != null ? DateFormat('dd MMM yyyy').format(invoice.dueDate!) : 'On Receipt'),
              _infoRow('Status:', invoice.paymentStatus),
            ])),
          ]),
          pw.SizedBox(height: 20),

          // Items Table
          pw.Table(
            columnWidths: {
              0: const pw.FlexColumnWidth(3),
              1: const pw.FixedColumnWidth(50),
              2: const pw.FixedColumnWidth(80),
              3: const pw.FixedColumnWidth(50),
              4: const pw.FixedColumnWidth(80),
            },
            children: [
              pw.TableRow(
                decoration: const pw.BoxDecoration(color: _green),
                children: ['Description', 'Qty', 'Unit Price', 'VAT %', 'Total'].map((h) =>
                  pw.Padding(padding: const pw.EdgeInsets.symmetric(horizontal: 8, vertical: 8),
                    child: pw.Text(h, style: pw.TextStyle(color: PdfColors.white, fontSize: 9, fontWeight: pw.FontWeight.bold),
                      textAlign: h == 'Description' ? pw.TextAlign.left : pw.TextAlign.right)),
                ).toList(),
              ),
              ...items.asMap().entries.map((entry) {
                final i = entry.key;
                final item = entry.value;
                return pw.TableRow(
                  decoration: pw.BoxDecoration(color: i.isEven ? PdfColors.white : _lightGray),
                  children: [
                    _cell(item.productName),
                    _cell(item.quantity.toString(), right: true),
                    _cell('$currency ${_fmt(fmt, item.price)}', right: true),
                    _cell('${item.taxRate.toStringAsFixed(0)}%', right: true),
                    _cell('$currency ${_fmt(fmt, item.total)}', right: true),
                  ],
                );
              }),
            ],
          ),
          pw.SizedBox(height: 20),

          // Totals
          pw.Row(mainAxisAlignment: pw.MainAxisAlignment.end, children: [
            pw.SizedBox(width: 200, child: pw.Column(children: [
              _totalRow('Total (excl. VAT):', '$currency ${_fmt(fmt, invoice.subtotal)}'),
              _totalRow('VAT ($vatRate%):', '$currency ${_fmt(fmt, invoice.taxAmount)}'),
              if (invoice.discount > 0)
                _totalRow('Discount (${invoice.discount.toStringAsFixed(0)}%):', '-$currency ${_fmt(fmt, invoice.subtotal * invoice.discount / 100)}'),
              pw.Divider(color: _green),
              _totalRow('Total (incl. VAT):', '$currency ${_fmt(fmt, invoice.total)}', bold: true),
              if (invoice.paidAmount > 0) ...[
                _totalRow('Paid:', '$currency ${_fmt(fmt, invoice.paidAmount)}'),
                _totalRow('Balance Due:', '$currency ${_fmt(fmt, invoice.balance)}', bold: true),
              ],
            ])),
          ]),

          if (invoice.isNote && invoice.noteReason != null) ...[
            pw.SizedBox(height: 16),
            pw.Text('Reason for issue:', style: pw.TextStyle(fontWeight: pw.FontWeight.bold, fontSize: 9)),
            pw.Text(invoice.noteReason!, style: const pw.TextStyle(fontSize: 9, color: PdfColors.grey700)),
          ] else if (invoice.notes != null) ...[
            pw.SizedBox(height: 16),
            pw.Text('Notes:', style: pw.TextStyle(fontWeight: pw.FontWeight.bold, fontSize: 9)),
            pw.Text(invoice.notes!, style: const pw.TextStyle(fontSize: 9, color: PdfColors.grey700)),
          ],

          // ZATCA QR code — the Base64 TLV payload issued with the invoice
          if (invoice.qrCode != null && invoice.qrCode!.isNotEmpty) ...[
            pw.SizedBox(height: 16),
            pw.Row(children: [
              pw.BarcodeWidget(
                barcode: bc.Barcode.qrCode(errorCorrectLevel: bc.BarcodeQRCorrectionLevel.medium),
                data: invoice.qrCode!,
                width: 80,
                height: 80,
                drawText: false,
              ),
              pw.SizedBox(width: 10),
              pw.Column(crossAxisAlignment: pw.CrossAxisAlignment.start, children: [
                pw.Text('Scan to verify — ZATCA e-invoice',
                    style: const pw.TextStyle(fontSize: 8, color: PdfColors.grey600)),
                _ar('امسح للتحقق من الفاتورة', size: 8, color: PdfColors.grey600),
              ]),
            ]),
          ],

          pw.Spacer(),

          // Footer
          pw.Container(
            padding: const pw.EdgeInsets.all(12),
            decoration: const pw.BoxDecoration(color: _green, borderRadius: pw.BorderRadius.all(pw.Radius.circular(8))),
            child: pw.Column(children: [
              pw.Text('$companyName  |  VAT $sellerVat  |  $companyAddress',
                  style: const pw.TextStyle(color: _whiteLight, fontSize: 8)),
              pw.SizedBox(height: 2),
              pw.Text('Electronically generated tax invoice issued under the KSA VAT Law and ZATCA e-invoicing regulations.',
                  style: const pw.TextStyle(color: _whiteLight, fontSize: 6)),
            ]),
          ),
        ],
      ),
    ));

    return doc.save();
  }

  pw.Widget _cell(String text, {bool right = false}) => pw.Padding(
    padding: const pw.EdgeInsets.symmetric(horizontal: 8, vertical: 6),
    child: pw.Text(text, style: const pw.TextStyle(fontSize: 9),
      textAlign: right ? pw.TextAlign.right : pw.TextAlign.left),
  );

  pw.Widget _infoRow(String label, String value) => pw.Padding(
    padding: const pw.EdgeInsets.symmetric(vertical: 2),
    child: pw.Row(mainAxisAlignment: pw.MainAxisAlignment.end, children: [
      pw.Text(label, style: const pw.TextStyle(color: PdfColors.grey600, fontSize: 9)),
      pw.SizedBox(width: 8),
      pw.Text(value, style: pw.TextStyle(fontSize: 9, fontWeight: pw.FontWeight.bold)),
    ]),
  );

  pw.Widget _totalRow(String label, String value, {bool bold = false}) => pw.Padding(
    padding: const pw.EdgeInsets.symmetric(vertical: 2),
    child: pw.Row(mainAxisAlignment: pw.MainAxisAlignment.spaceBetween, children: [
      pw.Text(label, style: pw.TextStyle(color: bold ? null : _gray, fontSize: bold ? 11 : 9, fontWeight: bold ? pw.FontWeight.bold : null)),
      pw.Text(value, style: pw.TextStyle(fontSize: bold ? 11 : 9, fontWeight: bold ? pw.FontWeight.bold : null,
        color: bold ? _maroon : null)),
    ]),
  );
}

File: backend/src/services/pdf.service.ts
import PDFDocument from 'pdfkit';
import path from 'path';
import fs from 'fs';
import QRCode from 'qrcode';
import * as zatca from './zatca.service';

const ArabicReshaper = require('arabic-reshaper');
const bidiFactory = require('bidi-js');
const bidi = bidiFactory();

/**
 * ZATCA-compliant tax invoice (فاتورة ضريبية) for Saudi Arabia.
 *
 * The layout carries every field ZATCA requires on a Phase 1 invoice: the
 * Arabic invoice title, seller name/VAT/address, buyer details, per-line VAT
 * breakdown, totals excluding and including VAT, and the Base64 TLV QR code.
 * Arabic is mandatory by law, so every label is printed bilingually.
 */

const FONT_DIR = path.join(__dirname, '../../assets/fonts');
const AR_REGULAR = path.join(FONT_DIR, 'Amiri-Regular.ttf');
const AR_BOLD = path.join(FONT_DIR, 'Amiri-Bold.ttf');
const hasArabicFont = fs.existsSync(AR_REGULAR) && fs.existsSync(AR_BOLD);

/**
 * PDFKit draws glyphs in logical order and does not shape Arabic, so text has
 * to be reshaped (contextual forms) and bidi-reordered before it is drawn.
 */
export const shapeArabic = (text: string): string => {
  if (!text) return '';
  try {
    const reshaped = ArabicReshaper.convertArabic(text);
    const embeddingLevels = bidi.getEmbeddingLevels(reshaped, 'rtl');
    const flips = bidi.getReorderSegments(reshaped, embeddingLevels);
    let chars = [...reshaped];
    for (const [start, end] of flips) {
      const segment = chars.slice(start, end + 1).reverse();
      chars = [...chars.slice(0, start), ...segment, ...chars.slice(end + 1)];
    }
    return chars.join('');
  } catch {
    return text;
  }
};

const AR = {
  taxInvoice: 'فاتورة ضريبية',
  simplifiedTaxInvoice: 'فاتورة ضريبية مبسطة',
  creditNote: 'إشعار دائن',
  debitNote: 'إشعار مدين',
  seller: 'المورد',
  buyer: 'العميل',
  vatNumber: 'الرقم الضريبي',
  crNumber: 'السجل التجاري',
  invoiceNo: 'رقم الفاتورة',
  issueDate: 'تاريخ الإصدار',
  supplyDate: 'تاريخ التوريد',
  dueDate: 'تاريخ الاستحقاق',
  description: 'الوصف',
  qty: 'الكمية',
  unitPrice: 'سعر الوحدة',
  vatRate: 'نسبة الضريبة',
  vatAmount: 'قيمة الضريبة',
  lineTotal: 'الإجمالي',
  subtotal: 'الإجمالي غير شامل الضريبة',
  vatTotal: 'إجمالي ضريبة القيمة المضافة',
  grandTotal: 'الإجمالي شامل الضريبة',
  paid: 'المبلغ المدفوع',
  balance: 'المبلغ المستحق',
  notes: 'ملاحظات',
  originalInvoice: 'الفاتورة الأصلية',
  reason: 'سبب الإصدار',
  zeroRated: 'خاضع لنسبة الصفر',
  exempt: 'معفى',
};

export interface InvoicePdfOptions {
  qrCode?: string | null;
}

export const generateInvoicePDF = async (
  invoice: any,
  settings: Record<string, string>,
  options: InvoicePdfOptions = {},
): Promise<Buffer> => {
  const currency = settings.currency || 'SAR';
  const sellerName = settings.company_name || 'Dubyx Trading Est.';
  const sellerNameAr = settings.company_name_ar || '';
  const sellerVat = settings.company_vat_number || '';
  const sellerCr = settings.company_cr_number || '';

  const isNote = invoice.invoiceKind && invoice.invoiceKind !== 'INVOICE';
  const isSimplified = invoice.invoiceType === 'SIMPLIFIED';

  // The QR is regenerated from stored invoice values so the PDF can never
  // disagree with the payload the customer scans.
  const qrPayload = options.qrCode || invoice.qrCode || zatca.generateQrCode({
    sellerName,
    vatNumber: sellerVat,
    timestamp: invoice.createdAt || new Date(),
    total: invoice.total || 0,
    vatTotal: invoice.taxAmount || 0,
  });
  const qrDataUrl = await QRCode.toDataURL(qrPayload, { margin: 0, width: 300, errorCorrectionLevel: 'M' });
  const qrBuffer = Buffer.from(qrDataUrl.split(',')[1], 'base64');

  return new Promise((resolve, reject) => {
    const doc = new PDFDocument({ size: 'A4', margin: 40 });
    const chunks: Buffer[] = [];
    doc.on('data', c => chunks.push(c));
    doc.on('end', () => resolve(Buffer.concat(chunks)));
    doc.on('error', reject);

    if (hasArabicFont) {
      doc.registerFont('ar', AR_REGULAR);
      doc.registerFont('ar-bold', AR_BOLD);
    }

    const green = '#1F6F4A';     // Saudi flag green
    const gold = '#C9A84C';
    const dark = '#1C1C1E';
    const gray = '#6B7280';
    const lightGray = '#F3F4F6';
    const n = (v: any) => (typeof v === 'number' && isFinite(v) ? v : 0);
    const money = (v: number) => `${currency} ${n(v).toFixed(2)}`;

    /** Right-aligned Arabic text; falls back silently when the font is absent. */
    const arabic = (text: string, x: number, y: number, width: number, size = 9, bold = false, color = dark) => {
      if (!hasArabicFont) return;
      doc.font(bold ? 'ar-bold' : 'ar').fontSize(size).fillColor(color)
        .text(shapeArabic(text), x, y, { width, align: 'right', features: ['liga', 'calt'] });
    };

    /* ── Header ── */
    doc.rect(0, 0, doc.page.width, 118).fill(green);

    doc.fillColor('white').font('Helvetica-Bold').fontSize(20).text(sellerName, 40, 26, { width: 300 });
    if (sellerNameAr) arabic(sellerNameAr, 40, 50, 300, 14, true, '#FFFFFF');

    doc.fillColor('white').font('Helvetica').fontSize(8)
      .text(`VAT No: ${sellerVat}`, 40, 74, { width: 300 })
      .text(`${sellerCr ? `CR No: ${sellerCr}  |  ` : ''}${settings.company_phone || ''}`, 40, 86, { width: 300 })
      .text(settings.company_address || '', 40, 98, { width: 300 });
    arabic(AR.vatNumber, 200, 73, 110, 8, false, '#FFFFFF');

    const titleEn = isNote
      ? (invoice.invoiceKind === 'CREDIT_NOTE' ? 'CREDIT NOTE' : 'DEBIT NOTE')
      : isSimplified ? 'SIMPLIFIED TAX INVOICE' : 'TAX INVOICE';
    const titleAr = isNote
      ? (invoice.invoiceKind === 'CREDIT_NOTE' ? AR.creditNote : AR.debitNote)
      : isSimplified ? AR.simplifiedTaxInvoice : AR.taxInvoice;

    doc.fillColor(gold).font('Helvetica-Bold').fontSize(15)
      .text(titleEn, doc.page.width - 260, 26, { width: 220, align: 'right' });
    arabic(titleAr, doc.page.width - 260, 46, 220, 16, true, '#FFFFFF');
    doc.fillColor('white').font('Helvetica').fontSize(10)
      .text(invoice.invoiceNumber, doc.page.width - 260, 74, { width: 220, align: 'right' });
    doc.fillColor('white').font('Helvetica').fontSize(7)
      .text(`UUID: ${invoice.uuid || ''}`, doc.page.width - 260, 90, { width: 220, align: 'right' });

    /* ── Parties + meta ── */
    const topY = 134;
    doc.fillColor(gray).font('Helvetica-Bold').fontSize(8).text('BILL TO', 40, topY);
    arabic(AR.buyer, 150, topY - 2, 60, 9, true, gray);

    const c = invoice.customer || {};
    doc.fillColor(dark).font('Helvetica-Bold').fontSize(11).text(c.shopName || '', 40, topY + 13, { width: 220 });
    if (c.arabicShopName) arabic(c.arabicShopName, 40, topY + 27, 220, 11, true);

    let by = topY + (c.arabicShopName ? 43 : 29);
    doc.fillColor(dark).font('Helvetica').fontSize(9);
    const buyerLines = [
      c.ownerName,
      c.phone,
      c.vatNumber ? `VAT No: ${c.vatNumber}` : null,
      c.crNumber ? `CR No: ${c.crNumber}` : null,
      [c.buildingNumber, c.street, c.district].filter(Boolean).join(', ') || c.address,
      [c.city, c.postalCode, c.additionalNumber].filter(Boolean).join(' '),
    ].filter(Boolean) as string[];
    buyerLines.forEach(line => { doc.text(line, 40, by, { width: 240 }); by += 12; });

    const metaX = 330;
    const drawMeta = (labelEn: string, labelAr: string, value: string, y: number) => {
      doc.fillColor(gray).font('Helvetica').fontSize(8).text(labelEn, metaX, y, { width: 95 });
      arabic(labelAr, metaX, y + 8, 95, 8, false, gray);
      doc.fillColor(dark).font('Helvetica-Bold').fontSize(9).text(value, metaX + 100, y + 2, { width: 125, align: 'right' });
    };
    const issued = new Date(invoice.createdAt || Date.now());
    drawMeta('Invoice No', AR.invoiceNo, invoice.invoiceNumber, topY);
    drawMeta('Issue Date', AR.issueDate, issued.toLocaleDateString('en-GB'), topY + 22);
    drawMeta('Supply Date', AR.supplyDate, new Date(invoice.supplyDate || issued).toLocaleDateString('en-GB'), topY + 44);
    drawMeta('Due Date', AR.dueDate, invoice.dueDate ? new Date(invoice.dueDate).toLocaleDateString('en-GB') : 'On Receipt', topY + 66);
    if (isNote && invoice.originalInvoice?.invoiceNumber) {
      drawMeta('Original Invoice', AR.originalInvoice, invoice.originalInvoice.invoiceNumber, topY + 88);
    }

    /* ── Items table ── */
    let tableY = topY + (isNote ? 118 : 100);
    const colX = [40, 250, 300, 375, 435, 495];
    const colW = [205, 45, 70, 55, 55, 60];
    const headersEn = ['Description', 'Qty', 'Unit Price', 'VAT %', 'VAT Amt', 'Total'];
    const headersAr = [AR.description, AR.qty, AR.unitPrice, AR.vatRate, AR.vatAmount, AR.lineTotal];

    const headerH = 30;
    doc.rect(40, tableY, doc.page.width - 80, headerH).fill(green);
    headersEn.forEach((h, i) => {
      doc.fillColor('white').font('Helvetica-Bold').fontSize(8)
        .text(h, colX[i], tableY + 4, { width: colW[i], align: i === 0 ? 'left' : 'right' });
      arabic(headersAr[i], colX[i], tableY + 16, colW[i], 7, false, '#FFFFFF');
    });

    let rowY = tableY + headerH;
    const items = invoice.lines || (invoice.order?.items || []).map((it: any) => {
      const category = it.product?.vatCategory || 'STANDARD';
      const rate = zatca.effectiveVatRate(category, n(it.taxRate));
      const lineTotal = n(it.price) * n(it.quantity) - n(it.discount);
      return {
        name: it.product?.name || '',
        nameAr: it.product?.nameAr || '',
        quantity: n(it.quantity),
        price: n(it.price),
        vatRate: rate,
        vatAmount: zatca.round2(lineTotal * rate / 100),
        lineTotal: zatca.round2(lineTotal),
        vatCategory: category,
      };
    });

    items.forEach((item: any, idx: number) => {
      if (rowY > doc.page.height - 260) { doc.addPage(); rowY = 50; }
      doc.rect(40, rowY, doc.page.width - 80, 20).fill(idx % 2 === 0 ? 'white' : lightGray);
      const suffix = item.vatCategory === 'ZERO_RATED' ? ' (Zero-rated)' : item.vatCategory === 'EXEMPT' ? ' (Exempt)' : '';
      doc.fillColor(dark).font('Helvetica').fontSize(8)
        .text(`${item.name}${suffix}`, colX[0], rowY + 6, { width: colW[0], ellipsis: true, height: 12 })
        .text(String(item.quantity), colX[1], rowY + 6, { width: colW[1], align: 'right' })
        .text(n(item.price).toFixed(2), colX[2], rowY + 6, { width: colW[2], align: 'right' })
        .text(`${n(item.vatRate).toFixed(0)}%`, colX[3], rowY + 6, { width: colW[3], align: 'right' })
        .text(n(item.vatAmount).toFixed(2), colX[4], rowY + 6, { width: colW[4], align: 'right' })
        .text(n(item.lineTotal + item.vatAmount).toFixed(2), colX[5], rowY + 6, { width: colW[5], align: 'right' });
      rowY += 20;
    });

    /* ── Totals ── */
    const totalsX = 330;
    let ty = rowY + 16;
    doc.moveTo(40, rowY + 8).lineTo(doc.page.width - 40, rowY + 8).strokeColor(lightGray).stroke();

    const drawTotal = (labelEn: string, labelAr: string, value: string, y: number, strong = false) => {
      doc.fillColor(strong ? dark : gray).font(strong ? 'Helvetica-Bold' : 'Helvetica').fontSize(strong ? 10 : 9)
        .text(labelEn, totalsX, y, { width: 100, align: 'right' });
      arabic(labelAr, totalsX - 120, y, 115, 9, strong, strong ? dark : gray);
      doc.fillColor(strong ? green : dark).font('Helvetica-Bold').fontSize(strong ? 12 : 9)
        .text(value, totalsX + 105, y - (strong ? 1 : 0), { width: 120, align: 'right' });
    };

    drawTotal('Total (excl. VAT)', AR.subtotal, money(invoice.subtotal), ty);
    ty += 16;
    if (n(invoice.discount) > 0) { drawTotal('Discount', 'الخصم', `-${money(invoice.discount)}`, ty); ty += 16; }
    drawTotal(`VAT (${settings.default_tax_rate || zatca.VAT_STANDARD_RATE}%)`, AR.vatTotal, money(invoice.taxAmount), ty);
    ty += 6;
    doc.rect(totalsX - 125, ty + 10, 355, 1).fill(green);
    ty += 16;
    drawTotal('Total (incl. VAT)', AR.grandTotal, money(invoice.total), ty, true);
    if (n(invoice.paidAmount) > 0) {
      ty += 20; drawTotal('Paid', AR.paid, money(invoice.paidAmount), ty);
      ty += 16; drawTotal('Balance Due', AR.balance, money(n(invoice.total) - n(invoice.paidAmount)), ty, true);
    }

    /* ── ZATCA QR code ── */
    const qrY = Math.min(ty + 30, doc.page.height - 190);
    doc.image(qrBuffer, 40, qrY, { width: 96, height: 96 });
    doc.fillColor(gray).font('Helvetica').fontSize(7)
      .text('Scan to verify — ZATCA e-invoice', 40, qrY + 100, { width: 110, align: 'center' });
    arabic('امسح للتحقق من الفاتورة', 30, qrY + 110, 130, 8, false, gray);

    if (isNote && invoice.noteReason) {
      doc.fillColor(gray).font('Helvetica-Bold').fontSize(8).text('Reason for issue:', 160, qrY);
      doc.fillColor(dark).font('Helvetica').fontSize(9).text(invoice.noteReason, 160, qrY + 12, { width: 250 });
    } else if (invoice.notes) {
      doc.fillColor(gray).font('Helvetica-Bold').fontSize(8).text('Notes:', 160, qrY);
      doc.fillColor(dark).font('Helvetica').fontSize(9).text(invoice.notes, 160, qrY + 12, { width: 250 });
    }

    /* ── Footer ── */
    // PDFKit starts a new page for any text drawn below the bottom margin, so the
    // margin is lifted while the footer band is painted.
    const savedBottomMargin = doc.page.margins.bottom;
    doc.page.margins.bottom = 0;
    const footerY = doc.page.height - 54;
    doc.rect(0, footerY, doc.page.width, 54).fill(green);
    doc.fillColor('white').font('Helvetica').fontSize(7)
      .text(
        `${sellerName}  |  VAT ${sellerVat}  |  ${settings.company_address || ''}`,
        40, footerY + 12, { align: 'center', width: doc.page.width - 80 },
      );
    doc.fillColor('#FFFFFFAA').font('Helvetica').fontSize(6)
      .text(
        'This is an electronically generated tax invoice issued under the KSA VAT Law and ZATCA e-invoicing regulations.',
        40, footerY + 26, { align: 'center', width: doc.page.width - 80 },
      );
    arabic('فاتورة ضريبية صادرة إلكترونياً وفقاً لنظام ضريبة القيمة المضافة وأحكام الفوترة الإلكترونية', 40, footerY + 36, doc.page.width - 80, 7, false, '#FFFFFF');
    doc.page.margins.bottom = savedBottomMargin;

    doc.end();
  });
};
